src/ray/rpc/retryable_grpc_client.h

```diff
@@ -0,0 +1,369 @@
+
+  /// This function is used to execute the given operation.
+  ///
+  /// \param operation The operation to be executed.
+  void Execute(std::function<void()> operation) {
+    operation_ = std::move(operation);
+    operation_();
+  }
+
+  /// This function is used to retry the given operation.
+  void Retry() { operation_(); }
+
+  void Abort(const ray::Status &status) { abort_callback_(status); }
+
+ private:
+  std::function<void(ray::Status)> abort_callback_;
+  std::function<void()> operation_;
+};
+
+/**
+ * The client makes RPC calls through the provided underlying grpc client.
+ * If the call goes through, the user provided callback is invoked.
+ * If the call fails due to transient network error, it is added to a retry queue.
+ * The client waits for the grpc channel reconnection to resend the requests.
+ * If the total number of request bytes in the queue exceeds max_pending_requests_bytes,
+ * the thread is blocked until some requests are resent.
+ * If a call's timeout_ms reaches during retry, its callback is called with
+ * Status::TimedOut. If the whole client does not reconnect within
+ * server_unavailable_timeout_seconds, server_unavailable_timeout_callback is invoked.
+ * When all callers of the client release the shared_ptr of the client, the client
+ * destructor is called and the client is shut down.
+ */
+class RetryableGrpcClient : public std::enable_shared_from_this<RetryableGrpcClient> {
+ public:
+  static std::shared_ptr<RetryableGrpcClient> Create(
+      std::shared_ptr<grpc::Channel> channel,
+      instrumented_io_context &io_context,
+      uint64_t max_pending_requests_bytes,
+      uint64_t check_channel_status_interval_milliseconds,
+      uint64_t server_unavailable_timeout_seconds,
+      std::function<void()> server_unavailable_timeout_callback,
+      const std::string &server_name) {
+    auto retryable_grpc_client = std::shared_ptr<RetryableGrpcClient>(
+        new RetryableGrpcClient(channel,
+                                io_context,
+                                max_pending_requests_bytes,
+                                check_channel_status_interval_milliseconds,
+                                server_unavailable_timeout_seconds,
+                                server_unavailable_timeout_callback,
+                                server_name));
+    return retryable_grpc_client;
+  }
+
+  RetryableGrpcClient(const RetryableGrpcClient &) = delete;
+  RetryableGrpcClient &operator=(const RetryableGrpcClient &) = delete;
+
+  template <typename Service, typename Request, typename Reply>
+  void CallMethod(PrepareAsyncFunction<Service, Request, Reply> prepare_async_function,
+                  GrpcClient<Service> &grpc_client,
+                  const std::string &call_name,
+                  const Request &request,
+                  const ClientCallback<Reply> &callback,
+                  const int64_t timeout_ms) {
+    auto executor = new Executor(
```

Comment:
I guess you use `new` and manage memory yourself probably due to `std::function` requires captured vars to be copy constructible?

Try https://github.com/abseil/abseil-cpp/blob/master/absl/functional/any_invocable.h?

Reply:
I generally use move only function whenever possible.
A few improvement compared with `std::function`, ref: https://github.com/facebook/folly/blob/main/folly/docs/Function.md

Reply:
If I'm guessing wrongly, I still prefer to use `std::shared_ptr`.
Managing memory manually is surprsing easy to get wrong.

Reply:
I believe it's not using `shared_ptr` to avoid circular dependency where `Executor` has a `shared_ptr` of itself via `operation_`

Reply:
Sorry I don't understand, whether using `new` or `shared_ptr` has no difference how object is constructed?

May I ask what's the compilation error you encountered?

Reply:
No compilation error. It's to prevent circular dependency caused memory leak:

```
class Hello {
  std::shared_ptr<Hello> self_;
}
```

Reply:
Ok, I managed to not use raw new and delete.